File: desktop/ai-app-bridge-cli/test/fixtures/p9-wikipedia-acceptance.py
def main(ctx):
    labels = ctx.inputs["labels"]
    call_ok(ctx, "wait-text", {"targetText": labels["home"], "timeoutSec": 8, "requireActivity": "MainActivity"})
    call_ok(ctx, "tap", {"tapX": 540, "tapY": 2256})
    call_ok(ctx, "tap-text", {"text": labels["searchField"]})
    call_ok(ctx, "wait-text", {"targetText": labels["back"], "timeoutSec": 8})
    call_ok(ctx, "input-text", {"text": labels["noResultQuery"], "tapX": 621, "tapY": 215})
    call_ok(ctx, "wait-text", {"targetText": labels["noResultQuery"], "timeoutSec": 8})
    call_ok(ctx, "hide-keyboard", {})
    call_ok(ctx, "wait-text", {"targetText": labels["noResult"], "timeoutSec": 15})
    empty = call_ok(ctx, "uia-tree", {"compact": True, "textFilter": labels["noResult"], "maxNodes": 16})
    empty_shot = call_ok(ctx, "screenshot", {})
    ctx.assert_({
        "name": "no-result",
        "predicateSummary": "frozen no-result label is in the compact uia tree",
        "condition": compact_has_text(empty, labels["noResult"]),
        "requiredEvidence": [],
        "requireCoverage": "complete",
        "evidence": empty_shot.get("evidence"),
    })
    call_ok(ctx, "tap-uia-text", {"text": labels["clearQuery"]})
    call_ok(ctx, "input-text", {"text": labels["articleTitle"], "tapX": 621, "tapY": 215})
    call_ok(ctx, "hide-keyboard", {})
    call_ok(ctx, "wait-text", {"targetText": labels["resultMarker"], "timeoutSec": 8})
    call_ok(ctx, "tap", {"tapX": 540, "tapY": 590})
    call_ok(ctx, "wait-text", {"targetText": labels["toc"], "timeoutSec": 8})
    article_title = call_ok(ctx, "uia-tree", {"compact": True, "textFilter": labels["articleTitle"], "maxNodes": 16})
    article_toc = call_ok(ctx, "uia-tree", {"compact": True, "textFilter": labels["toc"], "maxNodes": 16})
    article_shot = call_ok(ctx, "screenshot", {})
    ctx.assert_({
        "name": "article-visible",
        "predicateSummary": "frozen article title and TOC chrome are in the compact uia tree",
        "condition": compact_has_text(article_title, labels["articleTitle"]) and compact_has_text(article_toc, labels["toc"]),
        "requiredEvidence": [],
        "requireCoverage": "complete",
        "evidence": article_shot.get("evidence"),
    })
    call_ok(ctx, "tap-uia-text", {"text": labels["toc"]})
    call_ok(ctx, "wait-text", {"targetText": labels["tocSection"], "timeoutSec": 8})
    toc = call_ok(ctx, "tree", {})
    toc_shot = call_ok(ctx, "screenshot", {})
    ctx.assert_({
        "name": "toc-visible",
        "predicateSummary": "frozen TOC section is on the tree",
        "condition": labels["tocSection"] in collect_texts(toc.get("result")),
        "requiredEvidence": [],
        "requireCoverage": "complete",
        "evidence": toc_shot.get("evidence"),
    })
    call_ok(ctx, "tap", {"tapX": 573, "tapY": 570})
    call_ok(ctx, "wait-text", {"targetText": labels["readingList"], "timeoutSec": 8})
    call_ok(ctx, "tap-text", {"text": labels["readingList"]})
    listed = ctx.call("wait-text", {"targetText": labels["testList"], "timeoutSec": 4})
    if not listed or listed.get("ok") is False:
        call_ok(ctx, "wait-text", {"targetText": labels["readingListSheet"], "timeoutSec": 8})
        call_ok(ctx, "tap-uia-text", {"text": labels["readingListSheet"]})
        call_ok(ctx, "wait-text", {"targetText": labels["listNameField"], "timeoutSec": 8})
        call_ok(ctx, "input-text", {"text": labels["testList"]})
        call_ok(ctx, "tap-uia-text", {"text": labels["confirm"]})
        call_ok(ctx, "wait-text", {"targetText": labels["savedToList"], "timeoutSec": 8})
    else:
        call_ok(ctx, "tap-uia-text", {"text": labels["testList"]})
    call_ok(ctx, "wait-text", {"targetText": labels["savedToList"], "timeoutSec": 8})
    call_ok(ctx, "tap-text", {"text": labels["readingList"]})
    call_ok(ctx, "wait-text", {"targetText": labels["testList"], "timeoutSec": 8})
    test_list_dump = call_ok(ctx, "uia-tree", {"compact": True, "textFilter": labels["testList"], "maxNodes": 16})
    sheet_shot = call_ok(ctx, "screenshot", {})
    ctx.assert_({
        "name": "reading-list-sheet",
        "predicateSummary": "test collection p9-test is in the compact uia tree after opening the reading list",
        "condition": compact_has_text(test_list_dump, labels["testList"]),
        "requiredEvidence": [],
        "requireCoverage": "complete",
        "evidence": sheet_shot.get("evidence"),
    })
    call_ok(ctx, "tap-uia-text", {"text": labels["testList"]})
    call_ok(ctx, "wait-text", {"targetText": labels["readingList"], "timeoutSec": 8})
    call_ok(ctx, "tap-text", {"text": labels["theme"]})
    call_ok(ctx, "wait-text", {"targetText": labels["themeRestore"], "timeoutSec": 8})
    call_ok(ctx, "tap-uia-text", {"text": labels["themeRestore"]})
    call_ok(ctx, "tap", {"tapX": 315, "tapY": 1965})
    call_ok(ctx, "tap-uia-text", {"text": labels["themeRestore"]})
    call_ok(ctx, "tap", {"tapX": 84, "tapY": 216})
    call_ok(ctx, "wait-text", {"targetText": labels["language"], "timeoutSec": 8})
    call_ok(ctx, "tap-text", {"text": labels["language"]})
    call_ok(ctx, "wait-text", {"targetText": labels["languageLinksTitle"], "timeoutSec": 8})
    call_ok(ctx, "wait-text", {"targetText": labels["languagePage"], "timeoutSec": 60, "intervalMs": 3000})
    call_ok(ctx, "wait-text", {"targetText": labels["languageSwitch"], "timeoutSec": 10})
    call_ok(ctx, "tap-uia-text", {"text": labels["languageSwitch"], "exact": True})
    call_ok(ctx, "wait-text", {"targetText": labels["switchedTitle"], "timeoutSec": 60})
    call_ok(ctx, "wait-text", {"targetText": labels["toc"], "timeoutSec": 15})
    switched = call_ok(ctx, "uia-tree", {"compact": True, "textFilter": labels["switchedTitle"], "maxNodes": 16})
    switched_shot = call_ok(ctx, "screenshot", {})
    ctx.assert_({
        "name": "language-switched",
        "predicateSummary": "Afrikaans Moon title is in the compact uia tree after language switch",
        "condition": compact_has_text(switched, labels["switchedTitle"]),
        "requiredEvidence": [],
        "requireCoverage": "complete",
        "evidence": switched_shot.get("evidence"),
    })
    call_ok(ctx, "tap-uia-text", {"text": labels["back"]})
    call_ok(ctx, "wait-text", {
        "targetText": labels["articleTitle"],
        "absentText": labels["switchedTitle"],
        "timeoutSec": 60,
        "requireActivity": "PageActivity",
    })
    call_ok(ctx, "swipe", {"startX": 540, "startY": 400, "endX": 540, "endY": 1400, "durationMs": 400})
    call_ok(ctx, "swipe", {"startX": 540, "startY": 400, "endX": 540, "endY": 1400, "durationMs": 400})
    restored = call_ok(ctx, "uia-tree", {
        "compact": True,
        "textFilter": labels["articleTitle"],
        "visibleOnly": True,
        "maxNodes": 16,
    })
    restored_shot = call_ok(ctx, "screenshot", {})
    ctx.assert_({
        "name": "theme-language-restored",
        "predicateSummary": "visible Chinese article text is in compact uia nodes after theme cycle and back-stack language restore",
        "condition": compact_has_text(restored, labels["articleTitle"]),
        "requiredEvidence": [],
        "requireCoverage": "complete",
        "evidence": restored_shot.get("evidence"),
    })
    call_ok(ctx, "tap-uia-text", {"text": labels["back"]})
    call_ok(ctx, "wait-text", {"targetText": labels["resultMarker"], "timeoutSec": 8, "requireActivity": "SearchActivity"})
    left_article = call_ok(ctx, "uia-tree", {"compact": True, "textFilter": labels["resultMarker"], "maxNodes": 16})
    left_shot = call_ok(ctx, "screenshot", {})
    ctx.assert_({
        "name": "left-article-search",
        "predicateSummary": "search result marker is in the compact uia tree after leaving PageActivity",
        "condition": compact_has_text(left_article, labels["resultMarker"]),
        "requiredEvidence": [],
        "requireCoverage": "complete",
        "evidence": left_shot.get("evidence"),
    })
    call_ok(ctx, "tap-text", {"text": labels["back"]})
    call_ok(ctx, "wait-text", {"targetText": labels["home"], "timeoutSec": 8})
    call_ok(ctx, "tap-text", {"text": labels["more"]})
    call_ok(ctx, "wait-text", {"targetText": labels["settings"], "timeoutSec": 8})
    call_ok(ctx, "tap-text", {"text": labels["settings"]})
    call_ok(ctx, "wait-text", {"targetText": labels["settingsTheme"], "timeoutSec": 8})
    settings = call_ok(ctx, "tree", {})
    settings_shot = call_ok(ctx, "screenshot", {})
    ctx.assert_({
        "name": "settings-visible",
        "predicateSummary": "frozen settings theme row is on the tree",
        "condition": labels["settingsTheme"] in collect_texts(settings.get("result")),
        "requiredEvidence": [],
        "requireCoverage": "complete",
        "evidence": settings_shot.get("evidence"),
    })
    call_ok(ctx, "tap-text", {"text": labels["back"]})
    call_ok(ctx, "wait-text", {"targetText": labels["home"], "timeoutSec": 8})
    call_ok(ctx, "tap", {"tapX": 540, "tapY": 2256})
    call_ok(ctx, "tap-text", {"text": labels["searchField"]})
    call_ok(ctx, "input-text", {"text": labels["articleTitle"], "tapX": 621, "tapY": 215})
    call_ok(ctx, "hide-keyboard", {})
    call_ok(ctx, "wait-text", {"targetText": labels["resultMarker"], "timeoutSec": 8})
    call_ok(ctx, "tap", {"tapX": 540, "tapY": 590})
    call_ok(ctx, "wait-text", {"targetText": labels["toc"], "timeoutSec": 8})
    call_ok(ctx, "tap-uia-text", {"text": labels["back"]})
    call_ok(ctx, "wait-text", {"targetText": labels["resultMarker"], "timeoutSec": 8, "requireActivity": "SearchActivity"})
    repeat_left = call_ok(ctx, "uia-tree", {"compact": True, "textFilter": labels["resultMarker"], "maxNodes": 16})
    repeat_left_shot = call_ok(ctx, "screenshot", {})
    ctx.assert_({
        "name": "repeat-left-article-search",
        "predicateSummary": "search result marker is in the compact uia tree after the repeat article back",
        "condition": compact_has_text(repeat_left, labels["resultMarker"]),
        "requiredEvidence": [],
        "requireCoverage": "complete",
        "evidence": repeat_left_shot.get("evidence"),
    })
    ctx.checkpoint("wikipedia-acceptance", {"done": True})
    return {"passed": True}


def call_ok(ctx, command, args):
    result = ctx.call(command, args)
    if not result or result.get("ok") is False:
        raise RuntimeError(result.get("error") if result else command)
    return result


def collect_texts(tree):
    found = []
    if not isinstance(tree, dict):
        return found
    walk(tree.get("root"), found)
    return found


def compact_has_text(call_result, expected):
    result = call_result.get("result") if isinstance(call_result, dict) else None
    nodes = result.get("nodes", []) if isinstance(result, dict) else []
    return any(
        isinstance(value, str) and expected in value
        for node in nodes if isinstance(node, dict)
        for value in (node.get("text"), node.get("contentDescription"))
    )


def walk(node, found):
    if not node:
        return
    text = node.get("text")
    if isinstance(text, str):
        found.append(text)
    for child in node.get("children") or []:
        walk(child, found)

File: desktop/ai-app-bridge-cli/test/fixtures/p9-wikipedia-acceptance.js
'use strict';

async function main(ctx) {
  const labels = ctx.inputs.labels;
  await callOk(ctx, 'wait-text', { targetText: labels.home, timeoutSec: 8, requireActivity: 'MainActivity' });
  await callOk(ctx, 'tap', { tapX: 540, tapY: 2256 });
  await callOk(ctx, 'tap-text', { text: labels.searchField });
  await callOk(ctx, 'wait-text', { targetText: labels.back, timeoutSec: 8 });
  await callOk(ctx, 'input-text', { text: labels.noResultQuery, tapX: 621, tapY: 215 });
  await callOk(ctx, 'wait-text', { targetText: labels.noResultQuery, timeoutSec: 8 });
  await callOk(ctx, 'hide-keyboard', {});
  await callOk(ctx, 'wait-text', { targetText: labels.noResult, timeoutSec: 15 });
  const empty = await callOk(ctx, 'uia-tree', { compact: true, textFilter: labels.noResult, maxNodes: 16 });
  const emptyShot = await callOk(ctx, 'screenshot', {});
  await ctx.assert({
    name: 'no-result',
    predicateSummary: 'frozen no-result label is in the compact uia tree',
    condition: compactHasText(empty, labels.noResult),
    requiredEvidence: [],
    requireCoverage: 'complete',
    evidence: emptyShot.evidence,
  });
  await callOk(ctx, 'tap-uia-text', { text: labels.clearQuery });
  await callOk(ctx, 'input-text', { text: labels.articleTitle, tapX: 621, tapY: 215 });
  await callOk(ctx, 'hide-keyboard', {});
  await callOk(ctx, 'wait-text', { targetText: labels.resultMarker, timeoutSec: 8 });
  await callOk(ctx, 'tap', { tapX: 540, tapY: 590 });
  await callOk(ctx, 'wait-text', { targetText: labels.toc, timeoutSec: 8 });
  const articleTitle = await callOk(ctx, 'uia-tree', { compact: true, textFilter: labels.articleTitle, maxNodes: 16 });
  const articleToc = await callOk(ctx, 'uia-tree', { compact: true, textFilter: labels.toc, maxNodes: 16 });
  const articleShot = await callOk(ctx, 'screenshot', {});
  await ctx.assert({
    name: 'article-visible',
    predicateSummary: 'frozen article title and TOC chrome are in the compact uia tree',
    condition: compactHasText(articleTitle, labels.articleTitle) && compactHasText(articleToc, labels.toc),
    requiredEvidence: [],
    requireCoverage: 'complete',
    evidence: articleShot.evidence,
  });
  await callOk(ctx, 'tap-uia-text', { text: labels.toc });
  await callOk(ctx, 'wait-text', { targetText: labels.tocSection, timeoutSec: 8 });
  const toc = await callOk(ctx, 'tree', {});
  const tocShot = await callOk(ctx, 'screenshot', {});
  await ctx.assert({
    name: 'toc-visible',
    predicateSummary: 'frozen TOC section is on the tree',
    condition: collectTexts(toc.result).includes(labels.tocSection),
    requiredEvidence: [],
    requireCoverage: 'complete',
    evidence: tocShot.evidence,
  });
  await callOk(ctx, 'tap', { tapX: 573, tapY: 570 });
  await callOk(ctx, 'wait-text', { targetText: labels.readingList, timeoutSec: 8 });
  await callOk(ctx, 'tap-text', { text: labels.readingList });
  const listed = await ctx.call('wait-text', { targetText: labels.testList, timeoutSec: 4 });
  if (!listed || listed.ok === false) {
    await callOk(ctx, 'wait-text', { targetText: labels.readingListSheet, timeoutSec: 8 });
    await callOk(ctx, 'tap-uia-text', { text: labels.readingListSheet });
    await callOk(ctx, 'wait-text', { targetText: labels.listNameField, timeoutSec: 8 });
    await callOk(ctx, 'input-text', { text: labels.testList });
    await callOk(ctx, 'tap-uia-text', { text: labels.confirm });
    await callOk(ctx, 'wait-text', { targetText: labels.savedToList, timeoutSec: 8 });
  } else {
    await callOk(ctx, 'tap-uia-text', { text: labels.testList });
  }
  await callOk(ctx, 'wait-text', { targetText: labels.savedToList, timeoutSec: 8 });
  await callOk(ctx, 'tap-text', { text: labels.readingList });
  await callOk(ctx, 'wait-text', { targetText: labels.testList, timeoutSec: 8 });
  const testListDump = await callOk(ctx, 'uia-tree', { compact: true, textFilter: labels.testList, maxNodes: 16 });
  const sheetShot = await callOk(ctx, 'screenshot', {});
  await ctx.assert({
    name: 'reading-list-sheet',
    predicateSummary: 'test collection p9-test is in the compact uia tree after opening the reading list',
    condition: compactHasText(testListDump, labels.testList),
    requiredEvidence: [],
    requireCoverage: 'complete',
    evidence: sheetShot.evidence,
  });
  await callOk(ctx, 'tap-uia-text', { text: labels.testList });
  await callOk(ctx, 'wait-text', { targetText: labels.readingList, timeoutSec: 8 });
  await callOk(ctx, 'tap-text', { text: labels.theme });
  await callOk(ctx, 'wait-text', { targetText: labels.themeRestore, timeoutSec: 8 });
  await callOk(ctx, 'tap-uia-text', { text: labels.themeRestore });
  await callOk(ctx, 'tap', { tapX: 315, tapY: 1965 });
  await callOk(ctx, 'tap-uia-text', { text: labels.themeRestore });
  await callOk(ctx, 'tap', { tapX: 84, tapY: 216 });
  await callOk(ctx, 'wait-text', { targetText: labels.language, timeoutSec: 8 });
  await callOk(ctx, 'tap-text', { text: labels.language });
  await callOk(ctx, 'wait-text', { targetText: labels.languageLinksTitle, timeoutSec: 8 });
  await callOk(ctx, 'wait-text', { targetText: labels.languagePage, timeoutSec: 60, intervalMs: 3000 });
  await callOk(ctx, 'wait-text', { targetText: labels.languageSwitch, timeoutSec: 10 });
  await callOk(ctx, 'tap-uia-text', { text: labels.languageSwitch, exact: true });
  await callOk(ctx, 'wait-text', { targetText: labels.switchedTitle, timeoutSec: 60 });
  await callOk(ctx, 'wait-text', { targetText: labels.toc, timeoutSec: 15 });
  const switched = await callOk(ctx, 'uia-tree', { compact: true, textFilter: labels.switchedTitle, maxNodes: 16 });
  const switchedShot = await callOk(ctx, 'screenshot', {});
  await ctx.assert({
    name: 'language-switched',
    predicateSummary: 'Afrikaans Moon title is in the compact uia tree after language switch',
    condition: compactHasText(switched, labels.switchedTitle),
    requiredEvidence: [],
    requireCoverage: 'complete',
    evidence: switchedShot.evidence,
  });
  await callOk(ctx, 'tap-uia-text', { text: labels.back });
  await callOk(ctx, 'wait-text', {
    targetText: labels.articleTitle,
    absentText: labels.switchedTitle,
    timeoutSec: 60,
    requireActivity: 'PageActivity',
  });
  await callOk(ctx, 'swipe', { startX: 540, startY: 400, endX: 540, endY: 1400, durationMs: 400 });
  await callOk(ctx, 'swipe', { startX: 540, startY: 400, endX: 540, endY: 1400, durationMs: 400 });
  const restored = await callOk(ctx, 'uia-tree', {
    compact: true,
    textFilter: labels.articleTitle,
    visibleOnly: true,
    maxNodes: 16,
  });
  const restoredShot = await callOk(ctx, 'screenshot', {});
  await ctx.assert({
    name: 'theme-language-restored',
    predicateSummary: 'visible Chinese article text is in compact uia nodes after theme cycle and back-stack language restore',
    condition: compactHasText(restored, labels.articleTitle),
    requiredEvidence: [],
    requireCoverage: 'complete',
    evidence: restoredShot.evidence,
  });
  await callOk(ctx, 'tap-uia-text', { text: labels.back });
  await callOk(ctx, 'wait-text', { targetText: labels.resultMarker, timeoutSec: 8, requireActivity: 'SearchActivity' });
  const leftArticle = await callOk(ctx, 'uia-tree', { compact: true, textFilter: labels.resultMarker, maxNodes: 16 });
  const leftShot = await callOk(ctx, 'screenshot', {});
  await ctx.assert({
    name: 'left-article-search',
    predicateSummary: 'search result marker is in the compact uia tree after leaving PageActivity',
    condition: compactHasText(leftArticle, labels.resultMarker),
    requiredEvidence: [],
    requireCoverage: 'complete',
    evidence: leftShot.evidence,
  });
  await callOk(ctx, 'tap-text', { text: labels.back });
  await callOk(ctx, 'wait-text', { targetText: labels.home, timeoutSec: 8 });
  await callOk(ctx, 'tap-text', { text: labels.more });
  await callOk(ctx, 'wait-text', { targetText: labels.settings, timeoutSec: 8 });
  await callOk(ctx, 'tap-text', { text: labels.settings });
  await callOk(ctx, 'wait-text', { targetText: labels.settingsTheme, timeoutSec: 8 });
  const settings = await callOk(ctx, 'tree', {});
  const settingsShot = await callOk(ctx, 'screenshot', {});
  await ctx.assert({
    name: 'settings-visible',
    predicateSummary: 'frozen settings theme row is on the tree',
    condition: collectTexts(settings.result).includes(labels.settingsTheme),
    requiredEvidence: [],
    requireCoverage: 'complete',
    evidence: settingsShot.evidence,
  });
  await callOk(ctx, 'tap-text', { text: labels.back });
  await callOk(ctx, 'wait-text', { targetText: labels.home, timeoutSec: 8 });
  await callOk(ctx, 'tap', { tapX: 540, tapY: 2256 });
  await callOk(ctx, 'tap-text', { text: labels.searchField });
  await callOk(ctx, 'input-text', { text: labels.articleTitle, tapX: 621, tapY: 215 });
  await callOk(ctx, 'hide-keyboard', {});
  await callOk(ctx, 'wait-text', { targetText: labels.resultMarker, timeoutSec: 8 });
  await callOk(ctx, 'tap', { tapX: 540, tapY: 590 });
  await callOk(ctx, 'wait-text', { targetText: labels.toc, timeoutSec: 8 });
  await callOk(ctx, 'tap-uia-text', { text: labels.back });
  await callOk(ctx, 'wait-text', { targetText: labels.resultMarker, timeoutSec: 8, requireActivity: 'SearchActivity' });
  const repeatLeft = await callOk(ctx, 'uia-tree', { compact: true, textFilter: labels.resultMarker, maxNodes: 16 });
  const repeatLeftShot = await callOk(ctx, 'screenshot', {});
  await ctx.assert({
    name: 'repeat-left-article-search',
    predicateSummary: 'search result marker is in the compact uia tree after the repeat article back',
    condition: compactHasText(repeatLeft, labels.resultMarker),
    requiredEvidence: [],
    requireCoverage: 'complete',
    evidence: repeatLeftShot.evidence,
  });
  await ctx.checkpoint('wikipedia-acceptance', { done: true });
  return { passed: true };
}

async function callOk(ctx, command, args) {
  const result = await ctx.call(command, args);
  if (!result || result.ok === false) {
    throw new Error(result && result.error ? result.error : command);
  }
  return result;
}

function collectTexts(tree) {
  const found = [];
  walk(tree && tree.root, found);
  return found;
}

function compactHasText(callResult, expected) {
  const nodes = callResult && callResult.result && Array.isArray(callResult.result.nodes)
    ? callResult.result.nodes
    : [];
  return nodes.some((node) => [node.text, node.contentDescription]
    .some((value) => typeof value === 'string' && value.includes(expected)));
}

function walk(node, found) {
  if (!node) return;
  if (typeof node.text === 'string') found.push(node.text);
  for (const child of node.children || []) walk(child, found);
}

module.exports = { main };
